Expectations cljs (v2.1.0): Unable to resolve classname: js/Error

Hi @imikushin, just noticed something interesting while debugging the macro issue.

With the same [encore](https://github.com/ptaoussanis/encore) repo, using Cljs 0.0-2985:

Running `lein clean; lein cljx once; lein cljsbuild once main` works fine (Cljs is compiled, w/o problems).
Running `lein clean; lein cljx once; lein cljsbuild once tests` works fine (Cljs is compiled, w/o problems).
Running `lein clean; lein cljx once; lein cljsbuild once` produces an error:
```
module.js:338
    throw err;
          ^
Error: Cannot find module '/Users/ptaoussanis/Projects/encore/target/tests.js'
    at Function.Module._resolveFilename (module.js:336:15)
    at Function.Module._load (module.js:278:25)
    at Function.Module.runMain (module.js:501:10)
    at startup (node.js:129:16)
    at node.js:814:3
Reloading Clojure file "/taoensso/encore/tests.clj" failed.
clojure.lang.Compiler$CompilerException: java.lang.IllegalArgumentException: Unable to resolve classname: js/Error, compiling:(taoensso/encore/tests.clj:34:1)
             Compiler.java:6716 clojure.lang.Compiler.analyzeSeq
             Compiler.java:6500 clojure.lang.Compiler.analyze
             Compiler.java:6461 clojure.lang.Compiler.analyze
             Compiler.java:5837 clojure.lang.Compiler$BodyExpr$Parser.parse
             Compiler.java:5272 clojure.lang.Compiler$FnMethod.parse
             Compiler.java:3901 clojure.lang.Compiler$FnExpr.parse
             Compiler.java:6707 clojure.lang.Compiler.analyzeSeq
             Compiler.java:6500 clojure.lang.Compiler.analyze
             Compiler.java:6461 clojure.lang.Compiler.analyze
             Compiler.java:3713 clojure.lang.Compiler$InvokeExpr.parse
             Compiler.java:6711 clojure.lang.Compiler.analyzeSeq
             Compiler.java:6500 clojure.lang.Compiler.analyze
             Compiler.java:6461 clojure.lang.Compiler.analyze
             Compiler.java:2210 clojure.lang.Compiler$TryExpr$Parser.parse
             Compiler.java:6709 clojure.lang.Compiler.analyzeSeq
             Compiler.java:6500 clojure.lang.Compiler.analyze
               Compiler.java:38 clojure.lang.Compiler.access$200
             Compiler.java:6105 clojure.lang.Compiler$LetExpr$Parser.parse
             Compiler.java:6709 clojure.lang.Compiler.analyzeSeq
             Compiler.java:6500 clojure.lang.Compiler.analyze
             Compiler.java:6697 clojure.lang.Compiler.analyzeSeq
             Compiler.java:6500 clojure.lang.Compiler.analyze
             Compiler.java:6697 clojure.lang.Compiler.analyzeSeq
             Compiler.java:6500 clojure.lang.Compiler.analyze
             Compiler.java:6461 clojure.lang.Compiler.analyze
             Compiler.java:5837 clojure.lang.Compiler$BodyExpr$Parser.parse
             Compiler.java:5272 clojure.lang.Compiler$FnMethod.parse
             Compiler.java:3901 clojure.lang.Compiler$FnExpr.parse
             Compiler.java:6707 clojure.lang.Compiler.analyzeSeq
             Compiler.java:6500 clojure.lang.Compiler.analyze
             Compiler.java:6697 clojure.lang.Compiler.analyzeSeq
             Compiler.java:6500 clojure.lang.Compiler.analyze
               Compiler.java:38 clojure.lang.Compiler.access$200
              Compiler.java:560 clojure.lang.Compiler$DefExpr$Parser.parse
             Compiler.java:6709 clojure.lang.Compiler.analyzeSeq
             Compiler.java:6500 clojure.lang.Compiler.analyze
             Compiler.java:6461 clojure.lang.Compiler.analyze
             Compiler.java:6772 clojure.lang.Compiler.eval
             Compiler.java:7195 clojure.lang.Compiler.load
                    RT.java:370 clojure.lang.RT.loadResourceScript
                    RT.java:361 clojure.lang.RT.loadResourceScript
                    RT.java:440 clojure.lang.RT.load
                    RT.java:411 clojure.lang.RT.load
                  core.clj:5808 clojure.core/load[fn]
                  core.clj:5807 clojure.core/load
                RestFn.java:408 clojure.lang.RestFn.invoke
               compiler.clj:129 cljsbuild.compiler/reload-clojure[fn]
               compiler.clj:128 cljsbuild.compiler/reload-clojure
               compiler.clj:176 cljsbuild.compiler/run-compiler
form-init1797310464061109205.clj:1 user/eval3747[fn]
form-init1797310464061109205.clj:1 user/eval3747[fn]
                LazySeq.java:40 clojure.lang.LazySeq.sval
                LazySeq.java:49 clojure.lang.LazySeq.seq
                   Cons.java:39 clojure.lang.Cons.next
                LazySeq.java:81 clojure.lang.LazySeq.next
                    RT.java:644 clojure.lang.RT.next
                    core.clj:64 clojure.core/next
                  core.clj:3005 clojure.core/dorun
                  core.clj:3020 clojure.core/doall
form-init1797310464061109205.clj:1 user/eval3747
             Compiler.java:6768 clojure.lang.Compiler.eval
             Compiler.java:6758 clojure.lang.Compiler.eval
             Compiler.java:7195 clojure.lang.Compiler.load
             Compiler.java:7151 clojure.lang.Compiler.loadFile
                   main.clj:274 clojure.main/load-script
                   main.clj:279 clojure.main/init-opt
                   main.clj:307 clojure.main/initialize
                   main.clj:342 clojure.main/null-opt
                   main.clj:420 clojure.main/main
                RestFn.java:421 clojure.lang.RestFn.invoke
                   Var.java:383 clojure.lang.Var.invoke
                   AFn.java:156 clojure.lang.AFn.applyToHelper
                   Var.java:700 clojure.lang.Var.applyTo
                   main.java:37 clojure.main.main
Caused by: java.lang.IllegalArgumentException: Unable to resolve classname: js/Error
             Compiler.java:2247 clojure.lang.Compiler$TryExpr$Parser.parse
             Compiler.java:6709 clojure.lang.Compiler.analyzeSeq
```

Encore's project.clj :cljsbuild section has the "main" and "tests" profiles configured to compile **in parallel** when `lein cljsbuild once` is called without a specific profile name:
```clojure
   :builds
   [{:id "main"
     :source-paths   ["src" "target/classes"]
     ;; :notify-command ["terminal-notifier" "-title" "cljsbuild" "-m"]
     :compiler       {:output-to "target/main.js"
                      :optimizations :advanced
                      :pretty-print false}}
    {:id "tests"
     :source-paths   ["src" "target/classes" "test" "target/test-classes"]
     :notify-command ["node" "target/tests.js"]
     :compiler       {:output-to "target/tests.js"
                      :optimizations :simple ; Necessary for node.js
                      :pretty-print true
                      :target :nodejs
                      :hashbang false ; Ref. http://goo.gl/vrtNDR
                      :main "taoensso.encore.tests"}}]
```

So it seems that this error occurs only for **parallel builds**.

From what I can tell, the offender is `expectations.platform/err-type`:
```clojure
(defn err-type [] (if (cljs?) `~'js/Error `Throwable))
```

Looks like it might be some kind of race condition against `(defn cljs? [] (boolean (find-ns 'cljs.core)))`?

This isn't an important or urgent issue (it's easy to work around), just reporting it in case you weren't aware.

Cheers! :-)

Thanks, Peter. ClojureScript continues to amuze :) I'll definitely take a
look.

On Sat, Mar 21, 2015, 16:23 Peter Taoussanis <notifications@github.com>
wrote:

> Hi @imikushin <https://github.com/imikushin>, just noticed something
> interesting while debugging the macro issue.
>
> With the same encore <https://github.com/ptaoussanis/encore> repo, using
> Cljs 0.0-2985:
>
> Running lein clean; lein cljx once; lein cljsbuild once main works fine
> (Cljs is compiled, w/o problems).
> Running lein clean; lein cljx once; lein cljsbuild once tests works fine
> (Cljs is compiled, w/o problems).
> Running lein clean; lein cljx once; lein cljsbuild once produces an error:
>
> module.js:338
>     throw err;
>           ^
> Error: Cannot find module '/Users/ptaoussanis/Projects/encore/target/tests.js'
>     at Function.Module._resolveFilename (module.js:336:15)
>     at Function.Module._load (module.js:278:25)
>     at Function.Module.runMain (module.js:501:10)
>     at startup (node.js:129:16)
>     at node.js:814:3
> Reloading Clojure file "/taoensso/encore/tests.clj" failed.
> clojure.lang.Compiler$CompilerException: java.lang.IllegalArgumentException: Unable to resolve classname: js/Error, compiling:(taoensso/encore/tests.clj:34:1)
>              Compiler.java:6716 clojure.lang.Compiler.analyzeSeq
>              Compiler.java:6500 clojure.lang.Compiler.analyze
>              Compiler.java:6461 clojure.lang.Compiler.analyze
>              Compiler.java:5837 clojure.lang.Compiler$BodyExpr$Parser.parse
>              Compiler.java:5272 clojure.lang.Compiler$FnMethod.parse
>              Compiler.java:3901 clojure.lang.Compiler$FnExpr.parse
>              Compiler.java:6707 clojure.lang.Compiler.analyzeSeq
>              Compiler.java:6500 clojure.lang.Compiler.analyze
>              Compiler.java:6461 clojure.lang.Compiler.analyze
>              Compiler.java:3713 clojure.lang.Compiler$InvokeExpr.parse
>              Compiler.java:6711 clojure.lang.Compiler.analyzeSeq
>              Compiler.java:6500 clojure.lang.Compiler.analyze
>              Compiler.java:6461 clojure.lang.Compiler.analyze
>              Compiler.java:2210 clojure.lang.Compiler$TryExpr$Parser.parse
>              Compiler.java:6709 clojure.lang.Compiler.analyzeSeq
>              Compiler.java:6500 clojure.lang.Compiler.analyze
>                Compiler.java:38 clojure.lang.Compiler.access$200
>              Compiler.java:6105 clojure.lang.Compiler$LetExpr$Parser.parse
>              Compiler.java:6709 clojure.lang.Compiler.analyzeSeq
>              Compiler.java:6500 clojure.lang.Compiler.analyze
>              Compiler.java:6697 clojure.lang.Compiler.analyzeSeq
>              Compiler.java:6500 clojure.lang.Compiler.analyze
>              Compiler.java:6697 clojure.lang.Compiler.analyzeSeq
>              Compiler.java:6500 clojure.lang.Compiler.analyze
>              Compiler.java:6461 clojure.lang.Compiler.analyze
>              Compiler.java:5837 clojure.lang.Compiler$BodyExpr$Parser.parse
>              Compiler.java:5272 clojure.lang.Compiler$FnMethod.parse
>              Compiler.java:3901 clojure.lang.Compiler$FnExpr.parse
>              Compiler.java:6707 clojure.lang.Compiler.analyzeSeq
>              Compiler.java:6500 clojure.lang.Compiler.analyze
>              Compiler.java:6697 clojure.lang.Compiler.analyzeSeq
>              Compiler.java:6500 clojure.lang.Compiler.analyze
>                Compiler.java:38 clojure.lang.Compiler.access$200
>               Compiler.java:560 clojure.lang.Compiler$DefExpr$Parser.parse
>              Compiler.java:6709 clojure.lang.Compiler.analyzeSeq
>              Compiler.java:6500 clojure.lang.Compiler.analyze
>              Compiler.java:6461 clojure.lang.Compiler.analyze
>              Compiler.java:6772 clojure.lang.Compiler.eval
>              Compiler.java:7195 clojure.lang.Compiler.load
>                     RT.java:370 clojure.lang.RT.loadResourceScript
>                     RT.java:361 clojure.lang.RT.loadResourceScript
>                     RT.java:440 clojure.lang.RT.load
>                     RT.java:411 clojure.lang.RT.load
>                   core.clj:5808 clojure.core/load[fn]
>                   core.clj:5807 clojure.core/load
>                 RestFn.java:408 clojure.lang.RestFn.invoke
>                compiler.clj:129 cljsbuild.compiler/reload-clojure[fn]
>                compiler.clj:128 cljsbuild.compiler/reload-clojure
>                compiler.clj:176 cljsbuild.compiler/run-compiler
> form-init1797310464061109205.clj:1 user/eval3747[fn]
> form-init1797310464061109205.clj:1 user/eval3747[fn]
>                 LazySeq.java:40 clojure.lang.LazySeq.sval
>                 LazySeq.java:49 clojure.lang.LazySeq.seq
>                    Cons.java:39 clojure.lang.Cons.next
>                 LazySeq.java:81 clojure.lang.LazySeq.next
>                     RT.java:644 clojure.lang.RT.next
>                     core.clj:64 clojure.core/next
>                   core.clj:3005 clojure.core/dorun
>                   core.clj:3020 clojure.core/doall
> form-init1797310464061109205.clj:1 user/eval3747
>              Compiler.java:6768 clojure.lang.Compiler.eval
>              Compiler.java:6758 clojure.lang.Compiler.eval
>              Compiler.java:7195 clojure.lang.Compiler.load
>              Compiler.java:7151 clojure.lang.Compiler.loadFile
>                    main.clj:274 clojure.main/load-script
>                    main.clj:279 clojure.main/init-opt
>                    main.clj:307 clojure.main/initialize
>                    main.clj:342 clojure.main/null-opt
>                    main.clj:420 clojure.main/main
>                 RestFn.java:421 clojure.lang.RestFn.invoke
>                    Var.java:383 clojure.lang.Var.invoke
>                    AFn.java:156 clojure.lang.AFn.applyToHelper
>                    Var.java:700 clojure.lang.Var.applyTo
>                    main.java:37 clojure.main.main
> Caused by: java.lang.IllegalArgumentException: Unable to resolve classname: js/Error
>              Compiler.java:2247 clojure.lang.Compiler$TryExpr$Parser.parse
>              Compiler.java:6709 clojure.lang.Compiler.analyzeSeq
>
> Encore's project.clj :cljsbuild section has the "main" and "tests"
> profiles configured to compile *in parallel* when lein cljsbuild once is
> called without a specific profile name:
>
>    :builds
>    ;; TODO Parallel builds currently cause issues with Expectations v2.1.0,
>    ;; Ref.
>    [{:id "main"
>      :source-paths   ["src" "target/classes"]
>      ;; :notify-command ["terminal-notifier" "-title" "cljsbuild" "-m"]
>      :compiler       {:output-to "target/main.js"
>                       :optimizations :advanced
>                       :pretty-print false}}
>     {:id "tests"
>      :source-paths   ["src" "target/classes" "test" "target/test-classes"]
>      :notify-command ["node" "target/tests.js"]
>      :compiler       {:output-to "target/tests.js"
>                       :optimizations :simple ; Necessary for node.js
>                       :pretty-print true
>                       :target :nodejs
>                       :hashbang false ; Ref. http://goo.gl/vrtNDR
>                       :main "taoensso.encore.tests"}}]
>
> So it seems that this error occurs only for *parallel builds*.
>
> From what I can tell, the offender is expectations.platform/err-type:
>
> (defn err-type [] (if (cljs?) `~'js/Error `Throwable))
>
> Looks like it might be some kind of race condition against (defn cljs? []
> (boolean (find-ns 'cljs.core)))?
>
> This isn't an important or urgent issue (it's easy to work around), just
> reporting it in case you weren't aware.
>
> Cheers! :-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jaycfields/expectations/issues/54>.
>

--- a/project.clj
+++ b/project.clj
@@ -17,8 +17,8 @@
   :profiles {:dev {:dependencies [[org.clojure/clojure "1.8.0"]
                                   [org.clojure/clojurescript "1.8.34" :scope "provided"]]
                    :node-dependencies [[source-map-support "^0.2.9"]]
-                   :plugins           [[lein-cljsbuild "1.0.5"]
-                                       [lein-npm "0.5.0"]]}}
+                   :plugins           [[lein-cljsbuild "1.1.3"]
+                                       [lein-npm "0.6.2"]]}}
 
   :prep-tasks ["javac"]
   :auto-clean false

--- a/src/cljc/expectations/platform.cljc
+++ b/src/cljc/expectations/platform.cljc
@@ -7,8 +7,8 @@
   #?(:clj (:import (clojure.lang Agent Atom Ref))))
 
 #?(:clj
-   (defn cljs? []
-     (boolean (find-ns 'cljs.core))))
+   (defmacro cljs? []
+     (boolean (:ns &env))))
 
 #?(:clj
    (defn expanding [n]
@@ -32,9 +32,13 @@
   #?(:clj clojure.core/format
      :cljs goog.string/format))
 
+(defn nodejs? []
+  #?(:clj false
+     :cljs (= (js* "typeof(process)") "object")))
+
 (defn getenv [var]
   #?(:clj (System/getenv var)
-     :cljs (aget (.-env js/process) var)))
+     :cljs (aget (if (nodejs?) js/process.env js/window) var)))
 
 (defn get-message [e] (-> e
                           #?(:clj .getMessage
@@ -42,13 +46,16 @@
 
 (defn nano-time []
   #?(:clj (System/nanoTime)
-     :cljs (-> js/process .hrtime js->clj
-                (#(+ (* 1e9 (% 0)) (% 1))))))
+     :cljs (if (nodejs?)
+             (-> js/process .hrtime js->clj
+                 (#(+ (* 1e9 (% 0)) (% 1))))
+             (js/performance.now))))
+
 
 (defn on-windows? []
   (re-find #"[Ww]in"
            #?(:clj (System/getProperty "os.name")
-              :cljs (.-platform js/process))))
+              :cljs (if (nodejs?) (.-platform js/process) ""))))
 
 (def pprint
   #?(:clj pprint/pprint

